File: src/main/metadata.ts
import fetch from 'node-fetch'
import path from 'path'
import fs from 'fs'
import { app } from 'electron'
import { queries } from './database'

const cacheDir = () => path.join(app.getPath('userData'), 'cache', 'images')

function ensureCacheDir() {
  const dir = cacheDir()
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true })
}

async function downloadImage(url: string, filename: string): Promise<string | null> {
  try {
    ensureCacheDir()
    const localPath = path.join(cacheDir(), filename)
    if (fs.existsSync(localPath)) return localPath
    const response = await (fetch as any)(url, { timeout: 10000 })
    if (!response.ok) return null
    const buffer = await response.buffer()
    fs.writeFileSync(localPath, buffer)
    return localPath
  } catch { return null }
}

async function fetchSteamAppId(gameName: string): Promise<number | null> {
  try {
    const cacheKey = `steam_appid_${gameName.toLowerCase().replace(/\s+/g, '_')}`
    const cached = queries.getCacheEntry(cacheKey) as any
    if (cached?.metadata) {
      try { return JSON.parse(cached.metadata).appId || null } catch { return null }
    }
    const response = await (fetch as any)('https://api.steampowered.com/ISteamApps/GetAppList/v2/', { timeout: 15000 })
    if (!response.ok) return null
    const data: any = await response.json()
    const apps: any[] = data?.applist?.apps || []
    const normalized = gameName.toLowerCase().replace(/[^a-z0-9]/g, '')
    const match = apps.find((a: any) => {
      const n = (a.name || '').toLowerCase().replace(/[^a-z0-9]/g, '')
      return n === normalized || n.includes(normalized) || normalized.includes(n)
    })
    if (match) {
      queries.setCacheEntry(cacheKey, '', '', JSON.stringify({ appId: match.appid }))
      return match.appid
    }
    return null
  } catch { return null }
}

export async function fetchGameMetadata(gameId: number, gameName: string, executable: string): Promise<void> {
  try {
    const cacheKey = `metadata_${executable.toLowerCase().replace('.exe', '')}`
    if (queries.getCacheEntry(cacheKey)) return

    const appId = await fetchSteamAppId(gameName)
    if (!appId) {
      queries.setCacheEntry(cacheKey, '', '', JSON.stringify({ noData: true }))
      return
    }

    const [iconPath, coverPath] = await Promise.all([
      downloadImage(`https://cdn.cloudflare.steamstatic.com/steam/apps/${appId}/capsule_231x87.jpg`, `icon_${appId}.jpg`),
      downloadImage(`https://cdn.cloudflare.steamstatic.com/steam/apps/${appId}/library_600x900.jpg`, `cover_${appId}.jpg`)
    ])

    if (iconPath) queries.updateGame(gameId, { icon: iconPath })
    if (coverPath) queries.updateGame(gameId, { coverImage: coverPath })

    queries.setCacheEntry(cacheKey, '', iconPath || '', JSON.stringify({ appId }))
  } catch { }
}

export async function cleanOldCache(days: number): Promise<number> {
  return queries.cleanCache(days)
}

export function getImageAsBase64(imagePath: string): string | null {
  try {
    if (!imagePath || !fs.existsSync(imagePath)) return null
    const buffer = fs.readFileSync(imagePath)
    const ext = path.extname(imagePath).slice(1).replace('jpg', 'jpeg')
    return `data:image/${ext};base64,${buffer.toString('base64')}`
  } catch { return null }
}

File: backend/metadata.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

func getCacheDir() string {
	userDataPath := getUserDataPath()
	cacheDir := filepath.Join(userDataPath, "cache", "images")
	os.MkdirAll(cacheDir, 0755)
	return cacheDir
}

func downloadImage(url, filename string) (string, error) {
	localPath := filepath.Join(getCacheDir(), filename)

	// Return if already exists
	if _, err := os.Stat(localPath); err == nil {
		return localPath, nil
	}

	// Download image
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to download image: %d", resp.StatusCode)
	}

	// Write to file
	file, err := os.Create(localPath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	if _, err := io.Copy(file, resp.Body); err != nil {
		return "", err
	}

	return localPath, nil
}

type SteamAppListResponse struct {
	AppList struct {
		Apps []struct {
			AppID int    `json:"appid"`
			Name  string `json:"name"`
		} `json:"apps"`
	} `json:"applist"`
}

func fetchSteamAppID(gameName string) (int, error) {
	// Check cache first
	cacheKey := fmt.Sprintf("steam_appid_%s", strings.ToLower(strings.ReplaceAll(gameName, " ", "_")))
	if cached := queryCacheEntry(cacheKey); cached != nil {
		if metadata, ok := cached["metadata"].(string); ok && metadata != "" {
			var data map[string]interface{}
			if err := json.Unmarshal([]byte(metadata), &data); err == nil {
				if appID, ok := data["appId"].(float64); ok {
					return int(appID), nil
				}
			}
		}
	}

	// Fetch from Steam API
	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Get("https://api.steampowered.com/ISteamApps/GetAppList/v2/")
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return 0, fmt.Errorf("steam api error: %d", resp.StatusCode)
	}

	var steamResp SteamAppListResponse
	if err := json.NewDecoder(resp.Body).Decode(&steamResp); err != nil {
		return 0, err
	}

	// Normalize game name by removing non-alphanumeric characters
	normalizedName := ""
	for _, r := range gameName {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			normalizedName += string(r)
		} else if r >= 'A' && r <= 'Z' {
			normalizedName += string(r - 'A' + 'a')
		}
	}

	// Find best match
	for _, app := range steamResp.AppList.Apps {
		normalizedApp := ""
		for _, r := range app.Name {
			if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
				normalizedApp += string(r)
			} else if r >= 'A' && r <= 'Z' {
				normalizedApp += string(r - 'A' + 'a')
			}
		}

		if normalizedApp == normalizedName {
			// Cache the result
			metadataJSON, _ := json.Marshal(map[string]int{"appId": app.AppID})
			setCacheEntry(cacheKey, "", "", string(metadataJSON))
			return app.AppID, nil
		}
	}

	return 0, fmt.Errorf("game not found on steam")
}

func fetchGameMetadata(gameID int, gameName, executable string) error {
	cacheKey := fmt.Sprintf("metadata_%s", strings.TrimSuffix(strings.ToLower(executable), ".exe"))

	// Check if already cached
	if queryCacheEntry(cacheKey) != nil {
		return nil
	}

	// Try to fetch Steam App ID
	appID, err := fetchSteamAppID(gameName)
	if err != nil {
		// Mark as no data available
		metadataJSON, _ := json.Marshal(map[string]bool{"noData": true})
		setCacheEntry(cacheKey, "", "", string(metadataJSON))
		return nil
	}

	// Download images
	iconURL := fmt.Sprintf("https://cdn.cloudflare.steamstatic.com/steam/apps/%d/capsule_231x87.jpg", appID)
	coverURL := fmt.Sprintf("https://cdn.cloudflare.steamstatic.com/steam/apps/%d/library_600x900.jpg", appID)

	iconPath, _ := downloadImage(iconURL, fmt.Sprintf("icon_%d.jpg", appID))
	coverPath, _ := downloadImage(coverURL, fmt.Sprintf("cover_%d.jpg", appID))

	// Update game with image paths
	if iconPath != "" {
		updateGame(gameID, map[string]interface{}{"icon": iconPath})
	}
	if coverPath != "" {
		updateGame(gameID, map[string]interface{}{"coverImage": coverPath})
	}

	// Cache metadata
	metadataJSON, _ := json.Marshal(map[string]int{"appId": appID})
	setCacheEntry(cacheKey, "", iconPath, string(metadataJSON))

	return nil
}

func getSystemStats() map[string]interface{} {
	stats := make(map[string]interface{})
	stats["cpu"] = 0
	stats["ram"] = 0
	stats["ramUsedMB"] = 0
	stats["ramTotalMB"] = 0
	stats["gpu"] = 0
	stats["gpuName"] = "GPU"

	// Try to get CPU usage via wmic
	cmd := exec.Command("cmd", "/c", "wmic cpu get loadpercentage /value")
	output, err := cmd.Output()
	if err == nil {
		lines := strings.Split(string(output), "\n")
		for _, line := range lines {
			if strings.Contains(line, "LoadPercentage") {
				parts := strings.Split(line, "=")
				if len(parts) == 2 {
					var cpu int
					fmt.Sscanf(strings.TrimSpace(parts[1]), "%d", &cpu)
					stats["cpu"] = cpu
				}
			}
		}
	}

	// Try to get memory usage via wmic
	cmd = exec.Command("cmd", "/c", "wmic OS get FreePhysicalMemory,TotalVisibleMemorySize /value")
	output, err = cmd.Output()
	if err == nil {
		lines := strings.Split(string(output), "\n")
		var free, total int64

		for _, line := range lines {
			if strings.Contains(line, "FreePhysicalMemory") {
				parts := strings.Split(line, "=")
				if len(parts) == 2 {
					fmt.Sscanf(strings.TrimSpace(parts[1]), "%d", &free)
				}
			}
			if strings.Contains(line, "TotalVisibleMemorySize") {
				parts := strings.Split(line, "=")
				if len(parts) == 2 {
					fmt.Sscanf(strings.TrimSpace(parts[1]), "%d", &total)
				}
			}
		}

		if total > 0 {
			ramUsed := ((total - free) / total) * 100
			stats["ram"] = ramUsed
			stats["ramUsedMB"] = (total - free) / 1024
			stats["ramTotalMB"] = total / 1024
		}
	}

	return stats
}
